File: 1997-First Day Where You Have Been in All the Rooms.py
class Solution(object):
    def firstDayBeenInAllRooms(self, nextVisit):
        """
        :type nextVisit: List[int]
        :rtype: int
        """
# https://www.youtube.com/watch?v=GYyHxkGO4J0
# dp[i]: the number of days required to visit the ith room for the first time
        kMod = 1000000007
        N = len(nextVisit)
        dp = [0]*N
        
        for i in range(1, N):
            nxt = nextVisit[i-1]
            dp[i] = (dp[i-1] + 2 + (dp[i-1] - dp[nxt]))%kMod
        
        return dp[-1]

File: JavaScript/1997-First Day Where You Have Been in All the Rooms.js
/**
 * @param {number[]} nextVisit
 * @return {number}
 */
var firstDayBeenInAllRooms = function(nextVisit) {
    const kMod = 1000000007, N = nextVisit.length;
    let dp = new Array(N).fill(0);
    for (let i=1; i<N; i++){
        let nxt = nextVisit[i-1];
        dp[i] = dp[i-1] + 2 + (dp[i-1] - dp[nxt]);
        dp[i] = (dp[i] + kMod)%kMod;
    }
    return dp[N-1];
};